doc/developer/design/20230119_ir_normalization.md

```diff
@@ -0,0 +1,147 @@
+This is however not as important as bringing expressions to normal forms; we should do this even if there is some cost.
+
+Normalization should ideally not be harmful.
+If normalization undoes optimizations, we must pay careful attention to where we use it, to ensure optimization makes forward progress.
+
+Ideally normalization would only make changes that all parties agree are fundamentally uninteresting.
+
+## Non-Goals
+
+<!--
+// Enumerate potential goals that are explicitly out of scope for the project
+// ie. what could we do or what do we want to do in the future - but are not doing now
+-->
+
+Normalization is not meant to encompass all of optimization.
+
+We expect to have transformations that dramatically alter an AST bringing two structurally different expressions to equivalent representations.
+For example, the `Union` operators may have two inputs that can be canceled (because one is the `Negate` of the other);
+the transformation that removes cancelable pairs does not need to be part of normalization.
+
+In the future we may want to include some optimizing transformations that are sufficiently easy and obvious. 
+For the moment, we can rely on the fact that we plan to use normalization as part of the optimizer, and all optimizations will be run alongside normalization.
+
+## Description
+
+<!--
+// Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+// If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+I came up with the following components of syntactic normalization.
+These are meant to be structural changes to a `MirRelationExpr`: those that do not risk a change in meaning, by restricting themselves to very relatively simple actions on the shape of the AST.
+Whether a transform is "syntactic" or an action "relatively simple" is up for debate; I hope we'll know these when we see them.
+
+### Simplification
+
+Some operators are just more complicated forms of other operators.
+A `Reduce` with a constant argument can move that argument to a `Map`.
+A `Join` with one input could be a `Filter`.
+A `Join` with a constant argument *may* be able to express that argument as a `Map`. 
+
+Some of these skate close to optimizations.
+Turning an entire `Reduce` into a `Map` relies on the `group_key` being a unique key for the input collection.
+I think we should start judiciously here, especialy as the complexity can spiral.
+Again, we'll be running normalization alongside more sophisticated transforms that can perform more interesting optimizations.
+
+### Movement
+
+Many operators commute.
+When it is appropriate to do so, we should move commuting operators to canonical representations.
+For example, contiguous blocks of `Map`, `Filter`, and `Project` operators can be re-ordered to be `Map`s, `Filter`s, and then `Project`s.
```

Comment:
`CanonicalizeMfp` does this. Currently, it additionally performs MFP CSE (by calling `MapFilterProject::optimize()`), which made me think how that relates to normalization: Do we want to make it a part of normalization, or we don't care, or we consider it harmful as a normalization? I'm not sure, because I actually have an example where we would be happier if the normal form of MFP expressions would be "everything inlined" rather than "everything CSEd": Then  `LiteralConstraints` wouldn't need to do [this](https://github.com/MaterializeInc/materialize/blob/831d35bec5bbb7f1ff2f7bc48e66fb573582194c/src/transform/src/literal_constraints.rs#L436). And `PredicatePushdown::push_filters_through_map` is also relevant.